doc/sphinx/language/extensions/canonical.rst

```diff
@@ -41,12 +41,57 @@ in :ref:`canonicalstructures`; here only a simple example is given.
    This command supports the :attr:`local` attribute.  When used, the
    structure is canonical only within the :cmd:`Section` containing it.
 
-   Assume that :token:`qualid` denotes an object ``(Build_struct`` |c_1| … |c_n| ``)`` in the
-   structure :g:`struct` of which the fields are |x_1|, …, |x_n|.
-   Then, each time an equation of the form ``(``\ |x_i| ``_)`` |eq_beta_delta_iota_zeta| |c_i| has to be
+   Assume that :token:`qualid` denotes an object :n:`(Build_struct c__1 … c__n)` in the
+   structure :g:`struct` of which the fields are :n:`x__1, …, x__n`.
+   Then, each time an equation of the form :n:`(x__i _)` |eq_beta_delta_iota_zeta| :n:`c__i` has to be
    solved during the type checking process, :token:`qualid` is used as a solution.
-   Otherwise said, :token:`qualid` is canonically used to extend the field |c_i|
-   into a complete structure built on |c_i|.
+   Otherwise said, :token:`qualid` is canonically used to extend the field :n:`x__i`
+   into a complete structure built on :n:`c__i` when :n:`c__i` unifies with :n:`(x__i _)`.
+
+   The following kinds of terms are supported for the fields :n:`c__i` of :token:`qualid`:
+
+   * Constants (defined through :cmd:`Definition`, :cmd:`Inductive` and constructors thereof,
+     :cmd:`Record` and fields thereof, etc.) and section variables of an active section,
+     applied to zero or more arguments.
+   * Any :token:`sort`.
+   * Literal functions:  `fun … => …`.
+   * Literal, non-dependent function types, i.e. implications: `… -> …`.
+   * Variables bound in :token:`qualid`.
+
+   When searching for a canonical extension, only the head symbol of any existing instance's field :n:`c__i`
+   is considered.
+   We call this head symbol the *key* and we say ":token:`qualid` keys field :n:`x__i` to :n:`k`" when :n:`c__i`'s
+   head symbol is key :n:`k`.
+   Keys are the only piece of information that is used for canonical extension.
+   The keys corresponding to the kinds of terms listed above are:
+
+   * For constants and section variables, potentially applied to arguments:
+     the constant or variable itself, disregarding any arguments.
+   * For sorts: the sort itself.
+   * For literal functions: a disembodied function key denoted `fun _ => _`, disregarding both its domain
+     and body.
```

Comment:
```suggestion
   * For literal functions: skip the abstractions and use the key of its body.
```

Reply:
Should we mention casts here and in the list of supported terms above (given your suggested code change to strip them)?

Reply:
I don't know, it is not so clear a user should be made aware that "type annotations" are represented by actual nodes in the terms and that code that does not quotient pattern matching on this detail is likely buggy. Or at least, this is my understanding of the story. Dunno what is the mental model of an average non-dev Coq user, I hope he does not see `:` as an actual term constructor.

Reply:
Seems reasonable. I think the rest of the documentation also doesn't really mention casts as separate constructs (although I can't claim to be very familiar with the docs). There is definitely a discussion to be had here but this MR is not the right context for it. :)

Reply:
I cannot say I understand the context of this discussion, but note that we have this: https://coq.github.io/doc/master/refman/language/core/definitions.html#type-cast

Reply:
This is the syntax, so it is normal that `term : type` is there, since you can write it.
The question is if the doc here should be explicit in the fact that `(fun x => t   : type)` is considered a "literal function". IMO, `1` is a literal, and `1: nat` is a literal too. (see also my proposed change up there in the code)

Reply:
Also, the old code was failing the case `(fun x => (t : type))` as well, which is no doubt a literal function.

Reply:
@ppedrot would it be reasonable to have a `EConstr.king` which strips casts?

Reply:
We already have a constr version, no reason not to have an econstr version

Reply:
My feeling is that it should be "the default" and in only a few places, eg the type checker or the printer, you want to keep the cast.

Reply:
I tend to agree with that (with my limited experience of commonly forgetting to handle casts)